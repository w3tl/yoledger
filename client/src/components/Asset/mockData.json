[{
  "__typename": "Account",
  "id": "1",
  "name": "Bank Card",
  "balance": 0
}, {
  "__typename": "Account",
  "id": "2",
  "name": "Cash",
  "balance": 10
}, {
  "__typename": "Account",
  "id": "3",
  "name": "Bank Card 2",
  "balance": 150
}]
