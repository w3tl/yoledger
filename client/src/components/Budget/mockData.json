{
  "__typename": "BudgetPlan",
  "periods": ["2018-06-10", "2018-06-25", "2018-07-10", "2018-07-25", "2018-08-10", "2018-08-25"],
  "accounts": [{
    "account": {
      "__typename": "Account",
      "id": "1",
      "name": "Food"
    },
    "allocations": [{
      "__typename": "Allocation",
      "date": "2018-06-10",
      "amount": 10,
      "balance": 0
    }, {
      "__typename": "Allocation",
      "date": "2018-06-25",
      "amount": 12,
      "balance": 0
    }, {
      "__typename": "Allocation",
      "date": "2018-07-10",
      "amount": 14,
      "balance": 0
    }, {
      "__typename": "Allocation",
      "date": "2018-07-25",
      "amount": 18,
      "balance": 0
    }, {
      "__typename": "Allocation",
      "date": "2018-08-10",
      "amount": 7,
      "balance": 0
    }, {
      "__typename": "Allocation",
      "date": "2018-08-25",
      "amount": 6,
      "balance": 0
    }]
  }]
}
