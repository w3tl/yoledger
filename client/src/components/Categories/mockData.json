{
  "expenses": [{
    "id": "21",
    "name": "Food",
    "type": "EXPENSE"
  }, {
    "id": "22",
    "name": "Beer",
    "type": "EXPENSE"
  }, {
    "id": "23",
    "name": "Train",
    "type": "EXPENSE"
  }],
  "incomes": [{
    "id": "31",
    "name": "Salary",
    "type": "INCOME"
  }, {
    "id": "32",
    "name": "Debts",
    "type": "INCOME"
  }, {
    "id": "33",
    "name": "Cashback",
    "type": "INCOME"
  }]
}
