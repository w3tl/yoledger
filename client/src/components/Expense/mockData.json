[{
  "id": "21",
  "name": "Food"
}, {
  "id": "22",
  "name": "Beer"
}, {
  "id": "23",
  "name": "Train"
}]
