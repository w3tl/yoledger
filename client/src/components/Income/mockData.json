[{
  "id": "31",
  "name": "Salary"
}, {
  "id": "32",
  "name": "Debts"
}, {
  "id": "33",
  "name": "Dividends"
}]
