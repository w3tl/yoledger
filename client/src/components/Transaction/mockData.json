[{
  "__typename": "Transaction",
  "id": "1",
  "amount": 10,
  "source": {
    "__typename": "Account",
    "id": "2",
    "name": "Cash"
  },
  "destination": {
    "__typename": "Account",
    "id": "21",
    "name": "Food"
  },
  "date": "2018-01-01T00:00:00.000Z"
}, {
  "__typename": "Transaction",
  "id": "2",
  "amount": 10,
  "source": {
    "__typename": "Account",
    "id": "2",
    "name": "Cash"
  },
  "destination": {
    "__typename": "Account",
    "id": "1",
    "name": "Bank Card"
  },
  "date": "2018-04-01T00:00:00.000Z"
}, {
  "__typename": "Transaction",
  "id": "3",
  "amount": 10,
  "source": {
    "__typename": "Account",
    "id": "1",
    "name": "Bank Card"
  },
  "destination": {
    "__typename": "Account",
    "id": "21",
    "name": "Food"
  },
  "date": "2018-01-06T00:00:00.000Z"
}]
