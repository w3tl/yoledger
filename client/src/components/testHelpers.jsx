import React from 'react';
import { shallow, mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

export const withProvider = Component => props => (
  <MockedProvider {...props}>
    <Component />
  </MockedProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return fixedDate;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
