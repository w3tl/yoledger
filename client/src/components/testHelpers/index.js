import React from 'react';
import { shallow, mount } from 'enzyme';
import { ApolloProvider } from 'react-apollo';
import { ApolloClient } from 'apollo-client';
import { InMemoryCache } from 'apollo-cache-inmemory';
import { SchemaLink } from 'apollo-link-schema';
import { makeExecutableSchema } from 'graphql-tools';
import typeDefs from './typeDefs';
import resolvers from './resolvers';

export const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));

const executableSchema = makeExecutableSchema({
  typeDefs,
  resolvers,
});

const cache = new InMemoryCache();
const link = new SchemaLink({ schema: executableSchema });
export const client = new ApolloClient({
  link,
  cache,
});

export const withProvider = Component => props => (
  <ApolloProvider client={client}>
    <Component client={client} />
  </ApolloProvider>
);

export const testRenderWithoutError = (Component) => {
  it('renders without error', async () => {
    const wrapper = shallow(Component);
    await wait();
    wrapper.update();
    expect(wrapper).toMatchSnapshot();
  });
};

export const testLoadingState = (Component) => {
  it('should render loading state initially', () => {
    const wrapper = mount(Component);
    expect(wrapper.text()).toEqual('Loading...');
  });
};

export const testErrorUI = (Component) => {
  it('should show error UI', async () => {
    const wrapper = mount(Component);
    await wait(0); // wait for response
    expect(wrapper.update().text()).toContain('Error');
  });
};

// THNX: http://blog.blakesimpson.co.uk/read/98-how-to-stub-date-in-jest-tests-building-a-stubdate-helper
export const stubDate = (fixedDate) => {
  const date = new Date(fixedDate);
  let originalDate;

  beforeAll(() => {
    originalDate = Date;

    global.Date = class extends Date {
      constructor() {
        super();

        return date;
      }
    };
  });

  afterAll(() => {
    global.Date = originalDate;
  });
};
