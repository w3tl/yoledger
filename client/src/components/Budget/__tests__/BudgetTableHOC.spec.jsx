/* eslint-disable react/prop-types */
import React from 'react';
import { mount } from 'enzyme';
import RawBudgetTable from '../BudgetTable';
import HOC, { updateAfterCreate } from '../BudgetTableHOC';
import { accounts, budgets } from '../mockData';
import { withProvider, testLoadingState } from '../../testHelpers';

const wait = (ms = 0) => new Promise(resolve => setTimeout(resolve, ms));
const periods = [new Date('2018-06-10'), new Date('2018-06-25'), new Date('2018-07-10')];

const Row = ({ account, periods: cols }) => (
  <tr>
    <td>{account.name}</td>
    {cols.map(period => <td key={period}>Cell</td>)}
  </tr>
);

jest.mock('../BudgetTableFooter', () => ({ onCreate }) => (
  <tbody>
    <tr>
      <td>
        <button
          id="addAccount"
          type="button"
          onClick={() => onCreate({ account: 'new account' })}
        >
          Add
        </button>
      </td>
    </tr>
  </tbody>
));

const TableWithData = HOC(RawBudgetTable);

const variables = {
  dateStart: periods[0].toISOString(),
  dateEnd: periods[periods.length - 1].toISOString(),
};

const BudgetTable = withProvider(() => (
  <TableWithData
    row={Row}
    periods={periods}
    {...variables}
  />
));

describe('BudgetTableHOC', () => {
  testLoadingState(<BudgetTable />);

  describe('updateAfterCreate', () => {
    it('should add new account', () => {
      const writeQuery = jest.fn();
      const cache = {
        readQuery: jest.fn().mockReturnValue({
          budgets: {
            accounts: [accounts[0]],
          },
        }),
        writeQuery,
      };
      const newAccount = { id: 'new', name: 'New account' };
      const data = { upsertBudget: { budget: { ...budgets[0], account: newAccount } } };
      updateAfterCreate(variables)(cache, { data });
      expect(cache.readQuery.mock.calls[0][0].variables).toMatchSnapshot('readQuery');
      expect(writeQuery.mock.calls[0][0].data).toMatchSnapshot('writeQuery');
    });
  });

  it('should have props', async () => {
    const wrapper = mount(<BudgetTable />);
    await wait();
    wrapper.update();
    const table = wrapper.find('BudgetTable').first();
    expect(table.prop('accounts')).toHaveLength(accounts.length);
    expect(table.prop('periods')).toHaveLength(periods.length);
    expect(table.prop('onCreate')).toBeDefined();
  });
});
