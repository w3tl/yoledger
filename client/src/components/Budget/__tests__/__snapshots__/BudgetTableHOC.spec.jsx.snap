// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BudgetTableHOC updateAfterCreate should add new account: readQuery 1`] = `
Object {
  "dateEnd": "2018-07-10T00:00:00.000Z",
  "dateStart": "2018-06-10T00:00:00.000Z",
}
`;

exports[`BudgetTableHOC updateAfterCreate should add new account: writeQuery 1`] = `
Object {
  "budgets": Object {
    "accounts": Array [
      Object {
        "__typename": "Account",
        "id": "1",
        "name": "Food",
      },
      Object {
        "id": "new",
        "name": "New account",
      },
    ],
  },
}
`;
